android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/activity/ActivitySchemaCache.java

```diff
@@ -0,0 +1,156 @@
+package org.sagebionetworks.bridge.android.manager.activity;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.google.common.collect.Maps;
+
+import org.sagebionetworks.bridge.rest.model.Activity;
+import org.sagebionetworks.bridge.rest.model.ActivityType;
+import org.sagebionetworks.bridge.rest.model.CompoundActivity;
+import org.sagebionetworks.bridge.rest.model.SchemaReference;
+import org.sagebionetworks.bridge.rest.model.SurveyReference;
+import org.sagebionetworks.bridge.rest.model.TaskReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+/**
+ * A cache of mappings from activities' identifiers (tasks, surveys, compound activities) to
+ * (SchemaReference, SurveyReference). This is to help identify how an activity's return data should
+ * be formatted.
+ */
+public class ActivitySchemaCache {
+    private static final Logger LOG = LoggerFactory.getLogger(ActivitySchemaCache.class);
+    @NonNull
+    private final Map<String, SchemaReference> taskIdentifierToSchema;
```

Comment:
Your point about the confusing nature and complexity is a good one. I was trying to preserve the entities and denormalize, like how it is on the server side. I was not looking at the service code while writing this, but the code ended up looking quite similar. I'll think more on it, and describe more of the entities, if you have any ideas.

Activity points to one of Task or Survey or CompoundActivity
CompoundActivity has many Tasks and Surveys

Each Task/Survey/CompoundActivity can belong to many Activities

Each Task has many multiple revisions, Survey has many createdOns